File: python/Algorithm/Boj12100.py
UP = 0
DOWN = 1
LEFT = 2
RIGHT = 3
DIRECTIONS = [UP, DOWN, LEFT, RIGHT]


def dfs(N, game_map, current_round):
    if current_round == 5:
        return max(max(row) for row in game_map)
    max_value = 0
    for direction in DIRECTIONS:
        changed_map = [row[:] for row in game_map]
        if direction == UP:
            for c in range(N):
                mergeble_idx = 0
                for r in range(N):
                    if r != mergeble_idx and changed_map[r][c] != 0:
                        if changed_map[mergeble_idx][c] == 0:
                            changed_map[r][c], changed_map[mergeble_idx][c] = \
                                changed_map[mergeble_idx][c], changed_map[r][
                                    c]
                        else:
                            if changed_map[mergeble_idx][c] == changed_map[r][c]:
                                changed_map[mergeble_idx][c] *= 2
                                changed_map[r][c] = 0
                                mergeble_idx += 1
                            else:
                                mergeble_idx += 1
                                changed_map[r][c], changed_map[mergeble_idx][c] = \
                                    changed_map[mergeble_idx][c], \
                                        changed_map[r][c]
        elif direction == DOWN:
            for c in range(N):
                mergeble_idx = N - 1
                for r in range(N - 1, -1, -1):
                    if r != mergeble_idx and changed_map[r][c] != 0:
                        if changed_map[mergeble_idx][c] == 0:
                            changed_map[r][c], changed_map[mergeble_idx][c] = \
                                changed_map[mergeble_idx][c], changed_map[r][
                                    c]
                        else:
                            if changed_map[mergeble_idx][c] == changed_map[r][c]:
                                changed_map[mergeble_idx][c] *= 2
                                changed_map[r][c] = 0
                                mergeble_idx -= 1
                            else:
                                mergeble_idx -= 1
                                changed_map[r][c], changed_map[mergeble_idx][c] = \
                                    changed_map[mergeble_idx][c], \
                                        changed_map[r][c]
        elif direction == RIGHT:
            for r in range(N):
                mergeble_idx = N - 1
                for c in range(N - 1, -1, -1):
                    if c != mergeble_idx and changed_map[r][c] != 0:
                        if changed_map[r][mergeble_idx] == 0:
                            changed_map[r][c], changed_map[r][mergeble_idx] = changed_map[r][
                                mergeble_idx], changed_map[r][
                                c]
                        else:
                            if changed_map[r][mergeble_idx] == changed_map[r][c]:
                                changed_map[r][mergeble_idx] *= 2
                                changed_map[r][c] = 0
                                mergeble_idx -= 1
                            else:
                                mergeble_idx -= 1
                                changed_map[r][c], changed_map[r][mergeble_idx] = changed_map[r][
                                    mergeble_idx], \
                                    changed_map[r][c]
        else:
            for r in range(N):
                mergeble_idx = 0
                for c in range(N):
                    if c != mergeble_idx and changed_map[r][c] != 0:
                        if changed_map[r][mergeble_idx] == 0:
                            changed_map[r][c], changed_map[r][mergeble_idx] = changed_map[r][
                                mergeble_idx], changed_map[r][
                                c]
                        else:
                            if changed_map[r][mergeble_idx] == changed_map[r][c]:
                                changed_map[r][mergeble_idx] *= 2
                                changed_map[r][c] = 0
                                mergeble_idx += 1
                            else:
                                mergeble_idx += 1
                                changed_map[r][c], changed_map[r][mergeble_idx] = changed_map[r][
                                    mergeble_idx], \
                                    changed_map[r][c]
        if game_map != changed_map:
            max_value = max(max_value, dfs(N, changed_map, current_round + 1))
        else:
            max_value = max(max_value, max(max(row) for row in game_map))

    return max_value


def solution(N, game_map):
    return dfs(N, game_map[:], 0)


N = int(input())
game_map = []

for _ in range(N):
    game_map.append(list(map(int, input().split())))

print(solution(N, game_map))

File: dart/algorithm/bin/boj_12100.dart
import 'dart:io';
import 'dart:math';

enum Direction { up, down, right, left }

bool compare2DArray(List<List> array1, List<List> array2) {
  for (var r = 0; r < array1.length; r++) {
    for (var c = 0; c < array1.length; c++) {
      if (array1[r][c] != array2[r][c]) {
        return false;
      }
    }
  }
  return true;
}

int getMaxValueFrom2DArray(List<List> array) {
  return array
      .map((row) => row.reduce((e1, e2) => e1 > e2 ? e1 : e2))
      .reduce((e1, e2) => e1 > e2 ? e1 : e2);
}

void swap2DArrayValue(
    List<List> array, int v1Row, int v1Col, int v2Row, int v2Col) {
  int temp = array[v1Row][v1Col];
  array[v1Row][v1Col] = array[v2Row][v2Col];
  array[v2Row][v2Col] = temp;
}

int dfs(int n, List<List> gameMap, int currentRound) {
  if (currentRound == 5) {
    return getMaxValueFrom2DArray(gameMap);
  }

  var maxValue = 0;
  for (var direction in Direction.values) {
    var changedGameMap = gameMap.map((row) => List.from(row)).toList();
    switch (direction) {
      case Direction.up:
        for (var c = 0; c < n; c++) {
          var mergeableIdx = 0;
          for (var r = 0; r < n; r++) {
            if (mergeableIdx == r || changedGameMap[r][c] == 0) continue;

            if (changedGameMap[mergeableIdx][c] == 0) {
              swap2DArrayValue(changedGameMap, r, c, mergeableIdx, c);
              continue;
            }

            if (changedGameMap[mergeableIdx][c] == changedGameMap[r][c]) {
              changedGameMap[mergeableIdx][c] *= 2;
              changedGameMap[r][c] = 0;
            } else {
              swap2DArrayValue(changedGameMap, r, c, mergeableIdx + 1, c);
            }
            mergeableIdx += 1;
          }
        }
        break;
      case Direction.down:
        for (var c = 0; c < n; c++) {
          var mergeableIdx = n - 1;
          for (var r = n - 1; r >= 0; r--) {
            if (mergeableIdx == r || changedGameMap[r][c] == 0) continue;

            if (changedGameMap[mergeableIdx][c] == 0) {
              swap2DArrayValue(changedGameMap, r, c, mergeableIdx, c);
              continue;
            }

            if (changedGameMap[mergeableIdx][c] == changedGameMap[r][c]) {
              changedGameMap[mergeableIdx][c] *= 2;
              changedGameMap[r][c] = 0;
            } else {
              swap2DArrayValue(changedGameMap, r, c, mergeableIdx - 1, c);
            }
            mergeableIdx -= 1;
          }
        }
        break;
      case Direction.right:
        for (var r = 0; r < n; r++) {
          var mergeableIdx = n - 1;
          for (var c = n - 1; c >= 0; c--) {
            if (mergeableIdx == c || changedGameMap[r][c] == 0) continue;

            if (changedGameMap[r][mergeableIdx] == 0) {
              swap2DArrayValue(changedGameMap, r, c, r, mergeableIdx);
              continue;
            }

            if (changedGameMap[r][mergeableIdx] == changedGameMap[r][c]) {
              changedGameMap[r][mergeableIdx] *= 2;
              changedGameMap[r][c] = 0;
            } else {
              swap2DArrayValue(changedGameMap, r, c, r, mergeableIdx - 1);
            }
            mergeableIdx -= 1;
          }
        }
        break;
      case Direction.left:
        for (var r = 0; r < n; r++) {
          var mergeableIdx = 0;
          for (var c = 0; c < n; c++) {
            if (mergeableIdx == c || changedGameMap[r][c] == 0) continue;

            if (changedGameMap[r][mergeableIdx] == 0) {
              swap2DArrayValue(changedGameMap, r, c, r, mergeableIdx);
              continue;
            }

            if (changedGameMap[r][mergeableIdx] == changedGameMap[r][c]) {
              changedGameMap[r][mergeableIdx] *= 2;
              changedGameMap[r][c] = 0;
            } else {
              swap2DArrayValue(changedGameMap, r, c, r, mergeableIdx + 1);
            }
            mergeableIdx += 1;
          }
        }
        break;
    }

    if (compare2DArray(gameMap, changedGameMap)) {
      maxValue = max(maxValue, getMaxValueFrom2DArray(changedGameMap));
    } else {
      maxValue = max(maxValue, dfs(n, changedGameMap, currentRound + 1));
    }
  }

  return maxValue;
}

int solution(int n, List<List> gameMap) {
  return dfs(n, gameMap, 0);
}

void main() {
  var n = int.parse(stdin.readLineSync()!);

  List<List<int>> gameMap =
      List.generate(n, (index) => List.empty(growable: true));

  for (var r = 0; r < n; r++) {
    gameMap[r] = stdin.readLineSync()!.split(" ").map(int.parse).toList();
  }

  stdout.writeln(solution(n, gameMap));
}
